iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/twin/TwinPnPTests.java

```diff
@@ -0,0 +1,185 @@
+        isBasicTierHub = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_BASIC_TIER_HUB_ENV_VAR_NAME));
+        isPullRequest = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_PULL_REQUEST));
+
+        registryManager = RegistryManager.createFromConnectionString(iotHubConnectionString);
+
+        List inputs =  Arrays.asList(
+                new Object[][]
+                        {
+                                {MQTT},
+                                {MQTT_WS},
+                        }
+        );
+
+        return inputs;
+    }
+
+    private static class TestDevice
+    {
+        String deviceId;
+        String moduleId;
+        DeviceClient deviceClient;
+        ModuleClient moduleClient;
+    }
+
+    public TwinPnPTests.TwinPnPTestInstance testInstance;
+
+    public TwinPnPTests(IotHubClientProtocol protocol) throws IOException
+    {
+        this.testInstance = new TwinPnPTestInstance(protocol);
+    }
+
+    public class TwinPnPTestInstance
+    {
+        public IotHubClientProtocol protocol;
+        private com.microsoft.azure.sdk.iot.service.Device deviceForRegistryManager;
+        private com.microsoft.azure.sdk.iot.service.Module moduleForRegistryManager;
+
+        private DeviceTwin sCDeviceTwin;
+        private TwinPnPTests.TestDevice testDevice;
+
+        public TwinPnPTestInstance(IotHubClientProtocol protocol) throws IOException
+        {
+            this.protocol = protocol;
+            this.sCDeviceTwin = DeviceTwin.createFromConnectionString(iotHubConnectionString);
+            this.testDevice = new TestDevice();
+        }
+    }
+
+    @AfterClass
+    public static void tearDown() throws Exception
+    {
+        if (registryManager != null)
+        {
+            registryManager.close();
+            registryManager = null;
+        }
+    }
+
+    private void createDeviceWithClientOptions(IotHubClientProtocol protocol, ClientOptions options) throws IOException, URISyntaxException, InterruptedException
+    {
+        testInstance.testDevice.deviceClient = new DeviceClient(DeviceConnectionString.get(iotHubConnectionString, testInstance.deviceForRegistryManager), protocol, options);
+        testInstance.testDevice.deviceClient.open();
+    }
+
+    private void createModuleWithClientOptions(IotHubClientProtocol protocol, ClientOptions options) throws IOException, URISyntaxException, InterruptedException, ModuleClientException {
+        testInstance.testDevice.moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, testInstance.deviceForRegistryManager, testInstance.moduleForRegistryManager), protocol, options);
+        testInstance.testDevice.moduleClient.open();
+    }
+
+    @Before
+    public void createDevice() throws Exception
+    {
+        testInstance.testDevice = new TestDevice();
```

Comment:
Q - why do we have `new TestDevice()` here and in `TwinPnPTestInstance()` constructor as well?

Reply:
Good catch, removed the one from constructor